### PR TITLE
Sort intersect iterator's children iterators

### LINKED ISSUE #1880: Optimize intersect iterator - sort from low to high
Sort intersect iterator from a low to a high number of estimated results.
This will minimize the number of iteration we make.

### PR DESCRIPTION
**Pull request by ashtul**
Fixes #1880

### CODE REVIEW

File: src/index.c [hunk 1]
```diff
@@ -681,6 +681,15 @@ static void II_Rewind(void *ctx) {
   }
 }
 
+typedef int (*CompareFunc)(const void *a, const void *b);
+static int cmpIter(IndexIterator **it1, IndexIterator **it2) {
+  if (!*it1 && !*it2) return 0;
```
Comment (rafie) [inline src/index.c:686]: a little assert for `it1 && it2`?

File: tests/pytests/test_issues.py [hunk 2]
```diff
@@ -127,3 +127,30 @@ def test_issue1834(env):
...
+def test_issue1880(env):
+  # order of iterator in intersect is optimized by function
+  env.skipOnCluster()
+  conn = getConnectionByEnv(env)
+  env.cmd('FT.CONFIG', 'SET', '_PRINT_PROFILE_CLOCK', 'false')
+  conn.execute_command('FT.CREATE', 'idx', 'SCHEMA', 't', 'TEXT')
+  conn.execute_command('HSET', 'doc1', 't', 'hello world')
+  conn.execute_command('HSET', 'doc2', 't', 'hello')
+
+  excepted_res = [[1L, 'doc1', ['t', 'hello world']],
```
Comment (rafie) [inline tests/pytests/test_issues.py:140]: Not a change request, but this is exactly when JSON output of ft.profile will makes sense, so we don't need to rely on parsing user output.

File: src/index.c (hunk 1 above)
Comment (ashtul) [reply src/index.c:686]: These pointers are in the array passed to `qsort`. We might want to check that array earlier but here it cannot be NULL.

Comment (rafie) [reply src/index.c:686]: Are we completely sure the array cannot contain null elements? Dereferencing pointers without an assert causes me an itch.

Comment (ashtul) [reply src/index.c:686]: This is where it gets confusing :)
`qsort` returns a pointer to the type you have passed. In our case, we have passed an array of pointers.
It is legal to have null pointers, but in the compare function, we want to either push the null element to the end, and thus we return `1` or `-1` depends on the null value or return `0` if both are null.

File: src/index.c [hunk 3]
```diff
@@ -681,6 +681,15 @@ static void II_Rewind(void *ctx) {
...
+  if (!*it1) return -1;
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/index.c:687]: This compare function is for an intersect iterator. If we do have a NULL iterator, we want it in the front so we can return immediately with no results.
For example, if we have two iterators with 100k hits each and a NULL iterator, we could spend many iterations between these two (if none of their results match) before we finally reach the third/null iterator and just return EOF.

[the remark this answers is not in the document]
Comment (oshadmi) [reply src/index.c:687]: OK great, looks good, just this [comment](https://github.com/RediSearch/RediSearch/pull/1881#discussion_r584736721) saying "but in the compare function, we want to either push the null element to the **end**"
which is not aligned.

[the remark this answers is not in the document]
Comment (ashtul) [reply src/index.c:687]: You are correct.
I've corrected the previous comment.

**Review by oshadmi — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]
Thanks for the discussion on the additional cost of the qsort - since num of iterators is limited, the cost is low and the benefit is potentially big (inverted indexes may have many entries)

**Comment by MeirShpilraien** [comment]
Very nice @ashtul :+1:

**Pull request merged**